pilot/pkg/proxy/envoy/v2/ondemand.go

```diff
@@ -0,0 +1,438 @@
+package v2
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+
+	xdsapi "github.com/envoyproxy/go-control-plane/envoy/api/v2"
+	ads "github.com/envoyproxy/go-control-plane/envoy/service/discovery/v2"
+	"github.com/prometheus/client_golang/prometheus"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/peer"
+	"google.golang.org/grpc/status"
+	networking "istio.io/api/networking/v1alpha3"
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pilot/pkg/networking/util"
+)
+
+func newDeltaXdsConnection(peerAddr string, stream DeltaDiscoveryStream) *XdsConnection {
+	return &XdsConnection{
+		pushChannel:  make(chan *XdsEvent),
+		PeerAddr:     peerAddr,
+		Clusters:     []string{},
+		Connect:      time.Now(),
+		deltaStream:  stream,
+		LDSListeners: []*xdsapi.Listener{},
+		RouteConfigs: map[string]*xdsapi.RouteConfiguration{},
+	}
+}
+
+//DeltaDiscoveryStream represents the GRPC connection for any delta streams.
+type DeltaDiscoveryStream interface {
+	Send(*xdsapi.DeltaDiscoveryResponse) error
+	Recv() (*xdsapi.DeltaDiscoveryRequest, error)
+	grpc.ServerStream
+}
+
+// update the node associated with the connection, after receiving a a packet from envoy.
+func (s *DiscoveryServer) initDeltaConnectionNode(discReq *xdsapi.DeltaDiscoveryRequest, con *XdsConnection) error {
+	con.mu.RLock() // may not be needed - once per connection, but locking for consistency.
+	if con.modelNode != nil {
+		con.mu.RUnlock()
+		return nil // only need to init the node on first request in the stream
+	}
+	con.mu.RUnlock()
+
+	if discReq.Node == nil || discReq.Node.Id == "" {
+		return errors.New("missing node id")
+	}
+	nt, err := model.ParseServiceNodeWithMetadata(discReq.Node.Id, model.ParseMetadata(discReq.Node.Metadata))
+	if err != nil {
+		return err
+	}
+	// Update the config namespace associated with this proxy
+	nt.ConfigNamespace = model.GetProxyConfigNamespace(nt)
+	nt.Locality = discReq.Node.Locality
+
+	if err := nt.SetServiceInstances(s.Env); err != nil {
+		return err
+	}
+
+	if util.IsLocalityEmpty(nt.Locality) {
+		// Get the locality from the proxy's service instances.
+		// We expect all instances to have the same locality. So its enough to look at the first instance
+		if len(nt.ServiceInstances) > 0 {
+			nt.Locality = util.ConvertLocality(nt.ServiceInstances[0].GetLocality())
+		}
+	}
+
+	// Set the sidecarScope associated with this proxy if its a sidecar.
+	if nt.Type == model.SidecarProxy {
+		nt.SetSidecarScope(s.globalPushContext())
+	}
+
+	con.mu.Lock()
+	con.modelNode = nt
+	if con.ConID == "" {
+		// first request
+		con.ConID = connectionID(discReq.Node.Id)
+	}
+	con.mu.Unlock()
+
+	return nil
+}
+
+func receiveDeltaXdsThread(con *XdsConnection, reqChannel chan *xdsapi.DeltaDiscoveryRequest, errP *error) {
+	defer close(reqChannel) // indicates close of the remote side.
```

Comment:
Same - the interface you define above might be used to unify the different types of Send/Recv.

We'll want to eventually unify this with the handling of SDS - which currently has a different impl and had quite a few bugs.